src/goto-synthesizer/cegis_verifier.cpp

```diff
@@ -53,6 +53,55 @@ static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
   return false;
 }
 
+static const exprt &
+get_checked_pointer_from_null_pointer_check(const exprt &violation)
+{
+  // A NULL-pointer check is the negation of an equation between the checked
+  // pointer and a NULL pointer.
+  const exprt *equal_expr = &violation;
+
+  // Extract the equal expression in the violation.
+  while(equal_expr->id() != ID_equal)
+  {
+    INVARIANT(equal_expr->operands().size() > 0, " No equal expression found.");
+    equal_expr = &equal_expr->operands()[0];
+  }
+
+  const exprt *lhs_pointer = &equal_expr->operands()[0];
+  const exprt *rhs_pointer = &equal_expr->operands()[1];
+
+  while(lhs_pointer->type().id() != ID_pointer)
+  {
+    INVARIANT(
+      lhs_pointer->operands().size() > 0,
+      " No pointer found in lhs of NULL-pointer checks.");
+    lhs_pointer = &lhs_pointer->operands()[0];
+  }
+
+  while(rhs_pointer->type().id() != ID_pointer)
+  {
+    INVARIANT(
+      rhs_pointer->operands().size() > 0,
+      " No pointer found in lhs of NULL-pointer checks.");
+    rhs_pointer = &rhs_pointer->operands()[0];
+  }
```

Comment:
Would a non-pointer-type operand really be acceptable here?

Reply:
It is ```equal_exprt(pointer_object(p1), pointer_object(p2));``` where ```p1``` is a null pointer and ```p2``` is a pointer-type operand. I will make it more explicitly.